experimental/e2e/storage/har.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.har.Log.Entries = append(s.har.Log.Entries, &har.Entry{
+		StartedDateTime: s.currentTime().Format(time.RFC3339),
+		Time:            0.0,
+		Comment:         s.newUUID(),
+		Cache: &har.Cache{
+			Comment: "Not cached",
+		},
+		Timings: &har.Timings{
+			Send:    0.0,
+			Wait:    0.0,
+			Receive: 0.0,
+		},
+		Request: &har.Request{
+			Method:      req.Method,
+			HTTPVersion: req.Proto,
+			URL:         req.URL.String(),
+			Headers:     reqHeaders,
+			QueryString: queryString,
+			Cookies:     reqCookies,
+			BodySize:    int64(len(reqBody)),
+			PostData: &har.PostData{
+				MimeType: req.Header.Get("Content-Type"),
+				Text:     string(reqBody),
+			},
+		},
+		Response: &har.Response{
+			Status:      int64(res.StatusCode),
+			StatusText:  res.Status,
+			HTTPVersion: res.Proto,
+			Headers:     resHeaders,
+			HeadersSize: -1,
+			BodySize:    int64(len(resBody)),
+			Cookies:     resCookies,
+			RedirectURL: res.Header.Get("Location"),
+			Content: &har.Content{
+				Size:     int64(len(resBody)),
+				MimeType: res.Header.Get("Content-Type"),
+				Text:     string(resBody),
+			},
+		},
+	})
+}
+
+// Entries converts HAR entries to a slice of Entry (http.Request and http.Response pairs).
+func (s *HARStorage) Entries() []*Entry {
+	entries := make([]*Entry, len(s.har.Log.Entries))
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for i, e := range s.har.Log.Entries {
+		postData := ""
+		if e.Request.PostData != nil {
+			postData = e.Request.PostData.Text
+		}
+		req, err := http.NewRequest(e.Request.Method, e.Request.URL, nil)
+		if err != nil {
+			fmt.Println("Failed to create request", "err", err)
+			continue
+		}
+		req.Body = ioutil.NopCloser(strings.NewReader(postData))
+		req.ContentLength = e.Request.BodySize
+		req.Header = make(http.Header)
+		for _, header := range e.Request.Headers {
+			req.Header.Add(header.Name, header.Value)
+		}
+
+		bodyReq := req.Clone(context.Background())
+		bodyReq.Body = ioutil.NopCloser(strings.NewReader(postData))
+		res := &http.Response{
+			StatusCode:    int(e.Response.Status),
+			Status:        e.Response.StatusText,
+			Proto:         e.Response.HTTPVersion,
+			Header:        make(http.Header),
+			Body:          ioutil.NopCloser(strings.NewReader(e.Response.Content.Text)),
+			ContentLength: int64(len(e.Response.Content.Text)),
+			Request:       bodyReq,
+		}
+
+		for _, header := range e.Response.Headers {
+			res.Header.Add(header.Name, header.Value)
+		}
+
+		// use the HAR entry's comment field to store the ID of the entry
+		id := e.Comment
+		if id == "" {
+			id = s.newUUID()
+			e.Comment = id
+		}
+
+		entries[i] = &Entry{
+			ID:       id,
+			Request:  req,
+			Response: res,
+		}
+	}
+
+	return entries
+}
+
+// Delete removes the HAR entry with the given ID.
+func (s *HARStorage) Delete(id string) bool {
+	for i, e := range s.har.Log.Entries {
```

Comment:
Do you also need to make use of the Mutex here?

Reply:
good catch. i'll address this and the other things mentioned in a separate PR